### PR TITLE
[FLINK-8599] Improve the failure behavior of the FileInputFormat for …

### PR DESCRIPTION
**Pull request by ChengzhiZhao**
## What is the purpose of the change

This pull request is intent to improve the failure behavior of the ContinuousFileReader, currently if a bad file (for example, a different schema been dropped in this folder) came to the path and flink will do several retries.
However, since the file path persist in the checkpoint, when people tried to resume from external checkpoint, it threw the following error on no file been found and the process cannot move forward.

`java.io.IOException: Error opening the Input Split s3a://myfile [0,904]: No such file or directory: s3a://myfile`

The change is to check if the path exist before open the file, if error occurs and bad file removed, flink should resume the process and continue.

## Brief change log
- *Add a file exist check before open the file *

## Verifying this change
- *Manually verified the change by introduce a bad file while continuously monitoring the folder, after remove the bad file, the process continued.*

## Does this pull request potentially affect one of the following parts:
  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)
  - If yes, how is the feature documented? (not applicable)

### CODE REVIEW

File: flink-core/src/main/java/org/apache/flink/api/common/io/FileInputFormat.java [hunk 1]
```diff
@@ -691,6 +691,12 @@ public void open(FileInputSplit fileSplit) throws IOException {
 			LOG.debug("Opening input split " + fileSplit.getPath() + " [" + this.splitStart + "," + this.splitLength + "]");
 		}
 
+		if (!exists(fileSplit.getPath())) {
```
Comment (steveloughran) [inline flink-core/src/main/java/org/apache/flink/api/common/io/FileInputFormat.java]: you are doubling the number of checks for file existence here, which, when working with S3 implies three more HTTP requests which takes time and cost money. Better to do the open() call and catch FileNotFoundException, which all filesystems are required to throw if they are given a path which doesn't resolve to a file.

**Comment by steveloughran** [comment]
This is very inefficient against an object store, potentially adding a few hundred millis and $0.01 per file.  I would simply catch FileNotFoundExceptions raised in the open() call and treat them specially

See [How long does FileSystem.exists() take against S3?](http://steveloughran.blogspot.co.uk/2016/12/how-long-does-filesystemexists-take.html)

File: flink-core/src/main/java/org/apache/flink/api/common/io/FileInputFormat.java [hunk 2]
```diff
@@ -706,6 +700,9 @@ public void open(FileInputSplit fileSplit) throws IOException {
 			this.stream = isot.waitForCompletion();
 			this.stream = decorateInputStream(this.stream, fileSplit);
 		}
+		catch (FileNotFoundException e) {
+			throw new FileNotFoundException("Input split " + fileSplit.getPath() + " doesn't exist, skip and continue");
+		}
```
Comment (steveloughran) [inline flink-core/src/main/java/org/apache/flink/api/common/io/FileInputFormat.java]: I would recommend including the text and stack of the caught ex, for the better stack trace. FNFEs can get raised in odd circumstances in S3; seeing the full stack is what you need when fielding support calls. eg.

```java
throw new FileNotFoundException("Input split " + fileSplit.getPath() + " doesn't exist, skip and continue: " + e, e);
```

**Comment by ChengzhiZhao** [comment]
Thanks @steveloughran for your feedbacks, I updated based on your suggestions.

File: flink-core/src/main/java/org/apache/flink/api/common/io/FileInputFormat.java [hunk 3]
```diff
@@ -819,6 +819,10 @@ public void open(FileInputSplit fileSplit) throws IOException {
 			this.stream = isot.waitForCompletion();
 			this.stream = decorateInputStream(this.stream, fileSplit);
 		}
+		catch (FileNotFoundException e) {
+			throw new FileNotFoundException("Input split " + fileSplit.getPath() +
+					" doesn't exist, skip and continue: "  + e.getMessage());
+		}
```
Comment (steveloughran) [inline flink-core/src/main/java/org/apache/flink/api/common/io/FileInputFormat.java]: As this exception doesn't have a constructor which takes a nested exception, can you use `initCause()` to patch it.

```java
throw (FileNotFoundException)(new FileNotFoundException(...)).initCause(e)
```

**Comment by ChengzhiZhao** [comment]
@steveloughran Updated.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/functions/source/ContinuousFileReaderOperator.java [hunk 4]
```diff
@@ -340,6 +341,10 @@ public void run() {
 					}
 				}
 
+			} catch (FileNotFoundException e) {
+				if (LOG.isDebugEnabled()) {
```
Comment (StephanEwen) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/api/functions/source/ContinuousFileReaderOperator.java]: This needs to be more prominently logged than `debug`. Should be at least `info` or `warn`.

Please also use the placeholder syntax from log4j, to make the code simpler.

File: flink-core/src/main/java/org/apache/flink/api/common/io/FileInputFormat.java [hunk 5]
```diff
@@ -819,6 +819,10 @@ public void open(FileInputSplit fileSplit) throws IOException {
 			this.stream = isot.waitForCompletion();
 			this.stream = decorateInputStream(this.stream, fileSplit);
 		}
+		catch (FileNotFoundException e) {
+			throw (FileNotFoundException)(new FileNotFoundException("Input split " + fileSplit.getPath() +
```
Comment (StephanEwen) [inline flink-core/src/main/java/org/apache/flink/api/common/io/FileInputFormat.java]: I don't understand why "skip and continue" is in this message.
Not all users of the `FileIputFormat` skip and continue. The interpretation of the exception should not be assumed when creating the exception.

**Comment by kl0u** [comment]
@ChengzhiZhao I think that we should handle the "bad" splits better.
@StephanEwen You are right that we should give the user the option to fail the job or ignore them.

**Comment by ChengzhiZhao** [comment]
@StephanEwen @kl0u Thanks for you feedback, I will put an option for user to choose

**Comment by StephanEwen** [comment]
The approach here does not work (there is also no test, otherwise that would have been caught).
Flink does not magically load configs like Hadoop, Flink passes configs explicitly.

Aside from that, we first should actually decide whether this should be

**Comment by ChengzhiZhao** [comment]
@StephanEwen, thanks for your feedbacks. for our use case we'd like to take the bad file out to continue process instead of fixing the bad file to be able to continue. I think it can be more as job/application specific so we won't have it globally, it will give people flexibility. If that is the decision here, we can do as a connector parameter not via config which affect all the jobs. Please suggest.

**Comment by kl0u** [comment]
Hi @ChengzhiZhao , would you like to keep on working on the issue? The PR does not seem to have many conflicts with the master branch and with a bit of work it can get in.

If you want to work on this the missing bits are:
1) testing
2) remove the `ConfigOption` and add a connector parameter

If you do not want to work on the issue, please close the PR.
If you do not reply I will close it in the next couple of days.

**Comment by ChengzhiZhao** [comment]
Hi @kl0u, thanks for your feedback. Yes, I'd like to keep working on it. I'd try adding the missing parts and more testing for it.

**Comment by kl0u** [comment]
Thanks a lot @ChengzhiZhao !

**Comment by ChengzhiZhao** [comment]
@kl0u, In your comment `add a connector parameter`, do you mean add an additional parameter, for example `continueOnFileNotFound` on call of `createFileInput `

https://github.com/apache/flink/blob/15a7f978a2e591451c194c41408a12e64d04e9ba/flink-streaming-java/src/main/java/org/apache/flink/streaming/api/environment/StreamExecutionEnvironment.java#L1352-L1356

**Comment by kl0u** [comment]
I was thinking something like the `FileProcessingMode` in the `readFile` method.

**Comment by ChengzhiZhao** [comment]
Hey @kl0u, I made changes with the suggestions, please let me know your feedbacks. Thanks!

**Comment by ChengzhiZhao** [comment]
@kl0u , please let me know your feedbacks if we can merge the PR.

**Comment by aljoscha** [comment]
I'm closing this as "Abandoned", since there is no more activity and the code base has moved on quite a bit. Please re-open this if you feel otherwise and work should continue.

**Pull request closed without merging**